Add Joinable trait allowing filtering the results of one query with another

deontologician opened the pull request:
# Objective

There's a bit of boilerplate when using the entities from `Children` to join against a second query.

## Solution

Add a method on `Children` that takes a second query and calls `.get` on it in a loop, skipping entities that have no results in the subquery.

An alternative might be a more general `join` type of function that will subquery anything that is an `Iterator<Item=Entity>` and put that directly on `Query` maybe? Another option might be to add a `Join` at the type level.

alice-i-cecile commented:
CC @BoxyUwU @OptimisticPeach for a prototype of a relations feature using pure ECS.

alice-i-cecile commented:
FWIW, I think this is very useful functionality and should be merged in the meantime rather than waiting on relations. I'm a bit nervous about performance for this but I don't immediately see a faster approach.

mockersf commented on crates/bevy_transform/src/components/children.rs:
```diff
@@ -29,6 +31,33 @@ impl Children {
+    ///    children: Query<&Children, With<Sprite>>,
+    ///    transforms: Query<&Transform, With<Sprite>>,
+    /// ) {
+    ///     for transform in children.subquery(transforms) {
```
Wouldn't `children` here be a `Query` and not the `Children` component?

deontologician replied:
Oh, yep. That example needs an extra iteration. Good catch

mockersf commented on crates/bevy_transform/src/components/children.rs:
````diff
@@ -29,6 +31,33 @@ impl Children {
+    ///     }
+    /// }
+    /// ```
+    pub fn subquery<'a, Q, F>(
````
I would prefer `with_query` as name, as this is on a component, not on a query

deontologician replied:
Done

OptimisticPeach commented:
Unfortunately I have to agree with Alice in that this'll most likely be a bit slow, given the complexity of `query.get`. Additionally, it does not permit mutable queries.

I'd probably add a temporary `with_query_mut` which returns a streaming iterator of some sort to deal with this.

If it becomes a big issue prior to relations being merged (and they become powerful enough to do this), we might want to consider a faster solution built into `Query` for dealing with querying over `&[Entity]`.

OptimisticPeach commented:
Otherwise this isn't looking too bad, nice and elegant too \:P.

mockersf commented on crates/bevy_transform/src/components/children.rs:
```diff
@@ -29,6 +31,35 @@ impl Children {
         self.0.swap(a_index, b_index);
     }
+
+    /// Returns result of the subquery filtered by the children
```
This isn't returning `query.iter().filter(|entity| children.contains(entity))` as this comment would suggest, it returns an iterator over the children mapped and filtered on the query.

deontologician replied:
I rephrased it to

> Returns an iterator over the query result for each child

deontologician commented:
> I'd probably add a temporary `with_query_mut` which returns a streaming iterator of some sort to deal with this.

I took a quick stab at this, but it turned out to be kind of difficult to do so I gave up. I tried to implement a version where you pass in a callback that gets access to the mutable value, but that ended with an ICE: https://github.com/rust-lang/rust/issues/87531
:shrug:

alice-i-cecile commented:
@deontologician the ICE has since been fix; interested in taking another stab at this?

deontologician commented:
Yeah, I'll have some time later this week. I can give it a shot

deontologician commented:
Also, regarding design, I haven't kept up with how relations are going etc. Is this still a reasonable place to put this method?

cart commented:
Maybe this could be implemented for any Entity iterator?

then we could do:

```rust
for entity in some_entity_list.iter().with_query(some_query) {
} 
```

A variant like this might also be nice:
```rust
for (transform, velocity) in some_entity_list.iter().filter_map_query(some_query) {
}
```

This doesn't feel like it needs to be Children-specific functionality.

alice-i-cecile commented on crates/bevy_ecs/src/query/iter.rs:
```diff
@@ -477,3 +479,61 @@ where
     }
 }
+
+pub struct WithQuery<'world, 'state, I, Q, F>
```
Basic doc strings.

alice-i-cecile commented on crates/bevy_ecs/src/query/iter.rs:
```diff
@@ -477,3 +479,61 @@ where
+    }
+}
+
+pub trait WithQueryExt {
```
I don't like this naming convention. It's not really an extension trait: it's just a trait.

Not immediately sure on what to call it instead though.

deontologician replied:
I renamed it `Joinable`, not strongly attached to it though, so any suggestions welcome

alice-i-cecile commented on crates/bevy_ecs/src/query/iter.rs:
```diff
@@ -477,3 +479,61 @@ where
+    }
+}
+
+pub trait WithQueryExt {
```
This would be much clearer with a doc example :)

deontologician commented:
Fixed some formatting and renamed things a bit. I still need to add the doc comments and doc examples.

Also curious if this is the best location for the tests. I felt it was a little more natural in the system tests, even though it's defined in the query module.

alice-i-cecile commented:
I'm really liking the new direction. Could you update the PR title and description to reflect the changes? #1658 is probably worth mentioning.

> Also curious if this is the best location for the tests. I felt it was a little more natural in the system tests, even though it's defined in the query module.

Hmm. This feels like a query-joining tool; I'd prefer the tests stay with the query code.

alice-i-cecile commented:
@deontologician I'd like to see this design merged eventually, so I'm adding the `Adopt-Me` label. Feel free to remove it (or bug me) if you end up wanting to tackle this yourself.

mockersf commented:
is this case covered by [`Query::iter_many`](https://docs.rs/bevy/0.10.1/bevy/ecs/prelude/struct.Query.html#method.iter_many)?
